### PR TITLE
immortal: fixed formula to enable passing version on compile time

### PR DESCRIPTION
**Pull request by nbari**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/immortal.rb [hunk 1]
```diff
@@ -18,7 +18,22 @@ def install
     ENV["GOPATH"] = buildpath
     (buildpath/"src/github.com/immortal/immortal").install buildpath.children
     cd "src/github.com/immortal/immortal" do
-      system "make", "install", "DESTDIR=#{prefix}"
+      system "go", "get", "-u", "gopkg.in/yaml.v2"
```
Comment (DomT4) [inline Formula/immortal.rb]: Formulae should not run `go get` directly. You wanna look for `go_resource`s.

Comment (nbari) [inline Formula/immortal.rb]: Will take a look and update the request, thanks

Comment (nbari) [inline Formula/immortal.rb]: Should be fixed, could you please check.

File: Formula/immortal.rb [hunk 2]
```diff
@@ -14,11 +16,51 @@ class Immortal < Formula
...
     ENV["GOPATH"] = buildpath
     (buildpath/"src/github.com/immortal/immortal").install buildpath.children
+    Language::Go.stage_deps resources, buildpath/"src"
     cd "src/github.com/immortal/immortal" do
-      system "make", "install", "DESTDIR=#{prefix}"
+      ldflags = "-s -w -X main.version=#{version}"
+      system "go", "build", "-ldflags", ldflags, "-o", "immortal", "cmd/immortal/main.go"
+      system "go", "build", "-ldflags", ldflags, "-o", "immortalctl", "cmd/immortalctl/main.go"
+      system "go", "build", "-ldflags", ldflags, "-o", "immortaldir", "cmd/immortaldir/main.go"
+      bin.install "immortal"
```
Comment (DomT4) [inline Formula/immortal.rb]: This is _probably_ just about short enough to turn into an array:
```ruby
bin.install %w[immortal immortalctl immortaldir]
```

File: Formula/immortal.rb [hunk 3]
```diff
@@ -14,11 +16,51 @@ class Immortal < Formula
...
+      bin.install "immortalctl"
+      bin.install "immortaldir"
+      man8.install "man/immortal.8"
```
Comment (DomT4) [inline Formula/immortal.rb]: Are there any `.8` manpages we're not installing here? If we're installing everything anyway:
```ruby
man8.install Dir["man/*"]
```
or:
```ruby
man8.install Dir["man/*.8"]
```

**Review by DomT4 — COMMENTED** [review]
It's probably worth bumping the revision as well if this was missing resources & causing issues at run?

**Comment by DomT4** [comment]
> Regarding the version is there a way to dump/create a build version so that it won't interfere with the application itself?

I think what you're asking for is basically `revision`? Presuming you mean to force a new update without changing the upstream version or what version is passed to the build variable?

**Comment by nbari** [comment]
Exactly, how should I add a `revision`, could you please give me a hint?

should I just add it after `head`, for example:

```
class Immortal < Formula
  desc "OS agnostic (*nix) cross-platform supervisor"
  homepage "https://immortal.run/"
  url "https://github.com/immortal/immortal/archive/0.13.2.tar.gz"
  sha256 "f34674720828c984ef34df33b75b614fabe81e4fd50eb152746a33a273daa4f9"
  head "https://github.com/immortal/immortal.git"
  revision 2

   ...
```

**Comment by DomT4** [comment]
```diff
diff --git a/Formula/immortal.rb b/Formula/immortal.rb
index 9ee7daebf6..0dee21fc50 100644
--- a/Formula/immortal.rb
+++ b/Formula/immortal.rb
@@ -3,6 +3,8 @@ class Immortal < Formula
   homepage "https://immortal.run/"
   url "https://github.com/immortal/immortal/archive/0.13.2.tar.gz"
   sha256 "f34674720828c984ef34df33b75b614fabe81e4fd50eb152746a33a273daa4f9"
+  revision 1
+
   head "https://github.com/immortal/immortal.git"
 
   bottle do
```

**Comment by nbari** [comment]
Done, many thanks 👍

**Comment by DomT4** [comment]
@ilovezfs This is probably alright if you're happy. Guess what the `Makefile` was running! 😓😭

**Comment by nbari** [comment]
I like more this approach since looks/behave more like the FreeBSD port in terms that download the specific resources and don't delegate all to the Makefile, also for some reason the Makefile was not properly setting the version and seems that with this changes the version is now passed.

So from my side, I would like to keep this last version if is ok for you guys.

**Comment by DomT4** [comment]
Yeah, my comment was an endorsement of the change essentially.

Homebrew's policy is to not let projects run `go get` either themselves or from the formula, but since people don't have the time/memory/whatever to manually check every `Makefile` used in every `go` project sometimes these things slip through the net.

**Review by DomT4 — APPROVED** [review]

File: Formula/immortal.rb [hunk 4]
```diff
@@ -14,11 +17,47 @@ class Immortal < Formula
...
   def install
     ENV["GOPATH"] = buildpath
     (buildpath/"src/github.com/immortal/immortal").install buildpath.children
+    Language::Go.stage_deps resources, buildpath/"src"
     cd "src/github.com/immortal/immortal" do
-      system "make", "install", "DESTDIR=#{prefix}"
+      ldflags = "-s -w -X main.version=#{version}"
+      system "go", "build", "-ldflags", ldflags, "-o", "immortal", "cmd/immortal/main.go"
+      system "go", "build", "-ldflags", ldflags, "-o", "immortalctl", "cmd/immortalctl/main.go"
+      system "go", "build", "-ldflags", ldflags, "-o", "immortaldir", "cmd/immortaldir/main.go"
```
Comment (bfontaine) [inline Formula/immortal.rb]: Use `system "go", "build", "-ldflags", ldflags, "-o" "#{bin}/immortal", …` (same with other binaries) so you can remove the `bin.install` step below.

**Review by bfontaine — CHANGES_REQUESTED** [review]

File: Formula/immortal.rb (hunk 4 above)
Comment (nbari) [inline Formula/immortal.rb]: @bfontaine done, please check.

Comment (DomT4) [inline Formula/immortal.rb]: @bfontaine Can't believe I forgot that, nicely spotted!

Comment (nbari) [inline Formula/immortal.rb]: I accidentally,  pull force to my master, any idea how to recover the pull request?

**Comment by nbari** [comment]
Reopening after accidentally closing it

File: Formula/immortal.rb (hunk 4 above)
Comment (bfontaine) [inline Formula/immortal.rb]: What do you mean by "pull force to my master"? What command(s) did you run?

Comment (nbari) [inline Formula/immortal.rb]: I did a 

    git push nbari +master 

That removed the pull request, I just recover it but will try to clean all the commits, sorry for this inconvenience.

Comment (nbari) [inline Formula/immortal.rb]: @bfontaine Fixed, now before breaking it again, how to do a proper squash ? or can you do it when you merge the PR?

Thanks in advance.

Comment (nbari) [inline Formula/immortal.rb]: @bfontaine Hi, I was available to squash :) seems to be all now ok.

Comment (bfontaine) [inline Formula/immortal.rb]: Sorry I wasn’t in front of my computer but everything seems ok now :)

**Comment by nbari** [comment]
@bfontaine how to put in "green" the Changes requested option?

**Review by bfontaine — APPROVED** [review]

**Comment by bfontaine** [comment]
Done :)

**Pull request closed without merging**